File: check_file_permission.py
import minimatch
from pathlib import Path

def check_file_permission(file_path, operation, config):
    # Ensure file_path is a string for minimatch
    file_path_str = str(file_path)
    rules = config.get("filePermissions", [])
    for rule in rules:
        if minimatch.minimatch(file_path_str, rule["pattern"]):
            if operation in rule["operations"]:
                return rule["effect"] == "allow"
    return False  # Default-deny policy

File: ts-tools/check_file_permission.ts

import { minimatch } from 'minimatch';

export function checkFilePermission(filePath: string, operation: string, config: any): Promise<void> {
    return new Promise((resolve, reject) => {
        const rules = config.filePermissions || [];
        for (const rule of rules) {
            if (minimatch(filePath, rule.pattern)) {
                if (rule.operations.includes(operation)) {
                    if (rule.effect === "allow") {
                        resolve();
                        return;
                    }
                }
            }
        }
        reject(new Error(`Permission denied for ${operation} on ${filePath}`));
    });
}
